modules/dnn/src/onnx/onnx_importer.cpp

```diff
@@ -0,0 +1,348 @@
+    }
+
+    std::map<std::string, int> layer_id;
+    std::map<std::string, int>::iterator layerId, currentId;
+
+    int count = 0;
+    for (int j = 0; j < graph_proto.input_size(); j++)
+    {
+        if (constBlobs.find(graph_proto.input(j).name()) == constBlobs.end()) {
+            layer_id.insert(std::make_pair(graph_proto.input(j).name(), count));
+            count++;
+        }
+
+    }
+
+    int layersSize = graph_proto.node_size();
+    LayerParams layerParams;
+    opencv_onnx::NodeProto node_proto;
+
+    for(int i = 0; i < layersSize; i++)
+    {
+        node_proto = graph_proto.node(i);
+        layerParams = getLayerParams(node_proto);
+        CV_Assert(node_proto.output_size() >= 1);
+        layerParams.name = node_proto.output(0);
+
+        std::string layer_type = node_proto.op_type();
+        layerParams.type = layer_type;
+
+        if (layer_type == "MaxPool")
+        {
+            layerParams.type = "Pooling";
+            layerParams.set("pool", "MAX");
+            if (framework_name == "pytorch") {
+                layerParams.set("ceil_mode", false);   //  ceil_mode is not supported in export to ONNX in PyTorch
+            }
+        }
+        else if (layer_type == "AveragePool")
+        {
+            layerParams.type = "Pooling";
+            layerParams.set("pool", "AVE");
+            if (framework_name == "pytorch") {
+             layerParams.set("ceil_mode", false);   //  ceil_mode is not supported in export to ONNX in PyTorch
+            }
+        }
+        else if (layer_type == "GlobalAveragePool")
+        {
+            layerParams.type = "Pooling";
+            layerParams.set("pool", "AVE");
+            layerParams.set("global_pooling", true);
+        }
+        else if (layer_type == "LRN")
+        {
+            if (layerParams.has("size")) {
+                layerParams.set("local_size", layerParams.get<int>("size"));
+                layerParams.erase("size");
+            }
+        }
+        else if (layer_type == "BatchNormalization")
+        {
+            layerParams.type = "BatchNorm";
+            if (layerParams.has("epsilon")) {
+                layerParams.set("eps", layerParams.get<float>("epsilon"));
+                layerParams.erase("epsilon");
+            }
+            if (node_proto.input_size() != 5)
+                CV_Error(Error::StsNotImplemented,
+                         "Expected input, scale, bias, mean and var");
+
+            for (int j = 1; j < node_proto.input_size(); j++) {
+                layerParams.blobs.push_back( getBlob(node_proto, constBlobs, j) );
+            }
+            layerParams.set("scale_bias", true);
+        }
+        else if (layer_type == "Gemm")
+        {
+            layerParams.type = "InnerProduct";
+            std::map<std::string, Mat>::iterator constBlob;
+            for (int j = 0; j < node_proto.input_size(); j++) {
+                if (layer_id.find(node_proto.input(j)) == layer_id.end()) {
```

Comment:
Can we consider that Gemm layer always has inputs in the following order?
```
input
weights
bias (optional)
```
This way we can start a loop from 1 and remove a condition that input is trained parameters but not an output from previous layers.

Reply:
done